File: frontend/src/engine/capabilities/splitColumn.ts
/**
 * splitColumn – Split a text column into multiple columns by a delimiter.
 *
 * Common uses:
 *   - "First Last" → First Name / Last Name
 *   - "City, State" → City / State
 *   - "2024-01-15" → Year / Month / Day
 */

import { CapabilityMeta, SplitColumnParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange, resolveSheet, stripWorkbookQualifier } from "./rangeUtils";

const meta: CapabilityMeta = {
  action: "splitColumn",
  description: "Split a text column into multiple columns using a delimiter",
  mutates: true,
  affectsFormatting: false,
};

async function handler(
  context: Excel.RequestContext,
  params: SplitColumnParams,
  options: ExecutionOptions,
): Promise<StepResult> {
  const { sourceRange, delimiter, outputStartColumn, parts = 2 } = params;

  if (options.dryRun) {
    return { stepId: "", status: "success", message: `Would split ${sourceRange} by "${delimiter}" into ${parts} columns starting at ${outputStartColumn}` };
  }

  options.onProgress?.("Reading source column...");

  const ws = resolveSheet(context, sourceRange);
  const srcRng = resolveRange(context, sourceRange);
  const used = srcRng.getUsedRange(false);
  used.load("values, address");
  await context.sync();

  const vals = (used.values ?? []) as (string | number | boolean | null)[][];
  if (!vals.length) return { stepId: "", status: "success", message: "No data to split." };

  // Detect start row from address (e.g. "$A$2:$A$50" → 2)
  const cellPart = used.address.includes("!") ? used.address.split("!").pop()! : used.address;
  const startRow = parseInt((cellPart.replace(/\$/g, "").match(/[A-Z]+(\d+)/) ?? ["", "1"])[1], 10);

  options.onProgress?.(`Splitting ${vals.length} rows...`);

  const colToIndex = (col: string): number => {
    let n = 0;
    for (const c of col.toUpperCase()) n = n * 26 + (c.charCodeAt(0) - 64);
    return n - 1;
  };
  const indexToCol = (idx: number): string => {
    let s = "";
    idx++;
    while (idx > 0) { const r = (idx - 1) % 26; s = String.fromCharCode(65 + r) + s; idx = Math.floor((idx - 1) / 26); }
    return s;
  };

  const startColIdx = colToIndex(outputStartColumn);

  // Build the entire output grid in memory, then write as a single batch.
  // This avoids per-cell writes which are slow and can fail on merged cells.
  const outGrid: (string | number | boolean | null)[][] = [];

  for (let i = 0; i < vals.length; i++) {
    const cell = String(vals[i][0] ?? "");
    const splitParts = cell.split(delimiter).slice(0, parts);
    while (splitParts.length < parts) splitParts.push("");
    outGrid.push(splitParts.map((p) => p.trim()));
  }

  // Overwrite with headers if provided
  if (params.outputHeaders?.length && outGrid.length > 0) {
    for (let c = 0; c < Math.min(params.outputHeaders.length, parts); c++) {
      outGrid[0][c] = params.outputHeaders[c];
    }
  }

  // Single batch write — much faster and safer than per-cell writes
  const outAddr = `${indexToCol(startColIdx)}${startRow}:${indexToCol(startColIdx + parts - 1)}${startRow + outGrid.length - 1}`;
  try {
    ws.getRange(outAddr).values = outGrid as any;
    await context.sync();
  } catch (err: unknown) {
    const msg = err instanceof Error ? err.message : String(err);
    return { stepId: "", status: "error", message: `Failed to write split results: ${msg}. Range may contain merged or protected cells.` };
  }

  // Auto-fit the new columns
  const addr = stripWorkbookQualifier(sourceRange);
  const sheetPrefix = addr.includes("!") ? addr.split("!")[0] + "!" : "";
  try {
    const outRange = resolveRange(context, `${sheetPrefix}${outputStartColumn}${startRow}:${indexToCol(startColIdx + parts - 1)}${startRow + vals.length - 1}`);
    outRange.format.autofitColumns();
    await context.sync();
  } catch { /* non-fatal */ }

  return {
    stepId: "",
    status: "success",
    message: `Split ${vals.length} rows from ${sourceRange} into ${parts} columns starting at ${outputStartColumn}`,
  };
}

registry.register(meta, handler as any);
export { meta };

File: backend/app/execution/capabilities/split_column.py
"""splitColumn — split a text column into multiple columns by a delimiter.

Ported from `frontend/src/engine/capabilities/splitColumn.ts`. Uses the source
range's start row so the output aligns row-for-row, and writes a single 2D
batch so merged/protected cells fail loudly rather than silently partial.
"""

from __future__ import annotations

import re
from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import parse_address, resolve_range


def _col_letters_to_index(col: str) -> int:
    n = 0
    for c in col.upper():
        n = n * 26 + (ord(c) - 64)
    return n - 1


def _index_to_col_letters(idx: int) -> str:
    s = ""
    idx += 1
    while idx > 0:
        r = (idx - 1) % 26
        s = chr(65 + r) + s
        idx = (idx - 1) // 26
    return s


def _to_2d(values: Any) -> list[list[Any]]:
    if values is None:
        return []
    if not isinstance(values, list):
        return [[values]]
    if not values:
        return []
    if not isinstance(values[0], list):
        return [values]
    return values


def _start_row_from_address(addr: str) -> int:
    """Pull the first row number out of `Sheet!$A$2:$A$50` or similar."""
    cell_part = addr.split("!")[-1] if "!" in addr else addr
    cell_part = cell_part.replace("$", "")
    m = re.match(r"[A-Za-z]+(\d+)", cell_part)
    return int(m.group(1)) if m else 1


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    source_range = params.get("sourceRange")
    delimiter = params.get("delimiter")
    output_start_column = params.get("outputStartColumn")
    output_headers = params.get("outputHeaders")
    parts = int(params.get("parts") or 2)

    if not source_range or delimiter is None or not output_start_column:
        return {
            "status": "error",
            "message": "splitColumn requires 'sourceRange', 'delimiter', and 'outputStartColumn'.",
        }

    if ctx.dry_run:
        return {
            "status": "preview",
            "message": (
                f"Would split {source_range} by '{delimiter}' into {parts} columns "
                f"starting at {output_start_column}."
            ),
        }

    src = resolve_range(ctx.workbook_handle, source_range)
    vals = _to_2d(src.value)
    if not vals:
        return {"status": "success", "message": "No data to split.", "outputs": {"outputRange": None}}

    start_row = _start_row_from_address(src.address)

    out_grid: list[list[Any]] = []
    for row in vals:
        first_cell = row[0] if row else ""
        cell = "" if first_cell is None else str(first_cell)
        split_parts = cell.split(delimiter)[:parts]
        while len(split_parts) < parts:
            split_parts.append("")
        out_grid.append([p.strip() for p in split_parts])

    # Overwrite the first row with headers if provided (matches TS behavior).
    if output_headers and out_grid:
        for c in range(min(len(output_headers), parts)):
            out_grid[0][c] = output_headers[c]

    start_col_idx = _col_letters_to_index(output_start_column)
    end_col_letters = _index_to_col_letters(start_col_idx + parts - 1)
    start_col_letters = _index_to_col_letters(start_col_idx)

    # Keep output on the same sheet as the source range.
    parsed = parse_address(source_range)
    sheet_prefix = ""
    if parsed.sheet:
        needs_quote = any(ch in parsed.sheet for ch in " !'")
        sheet_prefix = f"'{parsed.sheet}'!" if needs_quote else f"{parsed.sheet}!"

    out_addr = (
        f"{sheet_prefix}{start_col_letters}{start_row}:"
        f"{end_col_letters}{start_row + len(out_grid) - 1}"
    )

    try:
        out_rng = resolve_range(ctx.workbook_handle, out_addr)
        out_rng.value = out_grid
    except Exception as exc:  # noqa: BLE001
        return {
            "status": "error",
            "message": (
                f"Failed to write split results: {exc}. Range may contain merged or protected cells."
            ),
            "error": str(exc),
        }

    # Best-effort autofit of the new columns (non-fatal).
    try:
        out_rng.columns.autofit()
    except Exception:  # noqa: BLE001
        pass

    return {
        "status": "success",
        "message": (
            f"Split {len(vals)} rows from {src.address} into {parts} columns "
            f"starting at {output_start_column}."
        ),
        "outputs": {"outputRange": out_rng.address},
    }


registry.register("splitColumn", handler, mutates=True)
